src/main/java/com/uid2/shared/secure/gcp/VmConfigId.java

```diff
@@ -0,0 +1,51 @@
+package com.uid2.shared.secure.gcp;
+
+public class VmConfigId {
+    private final String idString;
+    private final String failedReason;
+    private final boolean isValid;
+
+    // for troubleshooting purposes
+    private final String projectId;
+    public static VmConfigId success(String idString, String projectId) {
+        return new VmConfigId(true, idString, null, projectId);
+    }
+    public static VmConfigId failure(String reason, String projectId) {
+        return new VmConfigId(false, null, reason, projectId);
+    }
+
+    private VmConfigId(boolean success, String idString, String reason, String projectId) {
+        this.isValid = success;
+        this.idString = idString;
+        this.failedReason = reason;
+        this.projectId = projectId;
+    }
+
+    public boolean isValid() {
+        return this.isValid;
+    }
+
+    public String getValue() {
+        if (!isValid) throw new RuntimeException("value is invalid - check isValid() before using VmConfigId");
```

Comment:
Not sure of the java conventions, but throwing an exception in a getter could be very confusing for a caller. I would not expect a getter to ever throw an exception. Same for getFailedReason

Reply:
Comparing to a normal data type, this class is meant to carry the result of an attempt to generate VmConfigId (like Optional<String>), so the promoted usage is
```
final VmConfigId id = generateVmConfigId();
if(vmConfigId.isValid()) print(id.getValue());
else print(id.getFailedReason());
```
Calling the getter without checking validity is bad code and I want it to shout as loudly as possible :)

Reply:
I have had designs like this bite me hard in the past. What happens for example if the instance is passed to a logger that just calls the getters? What would be wrong with returning either an empty string, or a known value such as "IsValid not called, invalid result" or some other string?

Reply:
If it is used incorrectly and we returned null or invalid value, we are allowing invalid values to be passed into business logic and can cause wrong behaviors (for example, passing the attestation when it should not). The overriden "toString" method did reveal all the fields for logging purposes.

Reply:
Or put it another way: I want to make sure user checks "isValid" before using the idString, if there a compile-time way to enforce it I'm more than happy to do so 🤔

Reply:
Then can we change it to a method rather than a getter?

Reply:
I can't think of a proper name because the purpose of the method is getting an Optional value, and as Optional.get can throw NoSuchElementException I'm consider throwing fine.

Also I've changed it to assertion instead of runtimeException to indicate this is a logic error in code.

Reply:
removed the assertions. I think the problem I'm trying to solve is the caller ignores or forgets to check the return state, throwing is not the best way to deal with it but a static analysis tool (which is out of the scope of the change).